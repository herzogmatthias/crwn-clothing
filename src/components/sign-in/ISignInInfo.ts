export interface ISignInInfo {
  email: string;
  password: string;
}
