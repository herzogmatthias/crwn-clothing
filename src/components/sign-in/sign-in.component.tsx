import * as React from "react";
import { ISignInInfo } from "./ISignInInfo";
import FormInput from "../form-input/form-input.component";
import "./sign-in.styles.scss";
import { CustomButton } from "../custom-button/custom-button.component";
import { signInWithGoogle, auth } from "../../firebase/firebase.utils";

export interface ISignInProps {}

export function SignIn(props: ISignInProps) {
  const [signInInfo, setSignInInfo] = React.useState<ISignInInfo>({
    email: "",
    password: "",
  });
  const { email, password } = signInInfo;
  const _handleSubmit = async (event: React.FormEvent) => {
    event.preventDefault();
    try {
      await auth.signInWithEmailAndPassword(email, password);
      setSignInInfo({ email: "", password: "" });
    } catch (error) {
      console.log(error);
    }
  };

  const _handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { value, name } = event.target;
    setSignInInfo((prevState) => ({ ...prevState, [name]: value }));
  };

  return (
    <div className="sign-in">
      <h2>I already have an account</h2>
      <span>Sign in with your email and password</span>
      <form onSubmit={_handleSubmit}>
        <FormInput
          name="email"
          label="Email"
          type="email"
          value={email}
          handleChange={_handleChange}
          required
        ></FormInput>
        <FormInput
          label="Password"
          name="password"
          type="password"
          value={password}
          handleChange={_handleChange}
          required
        ></FormInput>
        <div className="buttons">
          <CustomButton type="submit">Sign In</CustomButton>
          <CustomButton isGoogleSignIn onClick={() => signInWithGoogle()}>
            Sign In with Google
          </CustomButton>
        </div>
      </form>
    </div>
  );
}
