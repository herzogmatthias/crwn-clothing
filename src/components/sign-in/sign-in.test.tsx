import React from "react";
import { mount, ReactWrapper } from "enzyme";
import { SignIn } from "./sign-in.component";
describe("NON CONNECTED SIGN-IN TESTS", () => {
  const mockGoogleSignInStart = jest.fn();
  const mockEmailSignInStart = jest.fn();
  const mockEmail = "asdf@gmail.com";
  const mockPassword = "asdf";
  let wrapper: ReactWrapper;
  const mockProps = {
    emailSignInStart: mockEmailSignInStart,
    googleSignInStart: mockGoogleSignInStart,
  } as any;

  beforeEach(() => {
    wrapper = mount(<SignIn {...mockProps}></SignIn>);
  });
  test("render SignIn Component", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("start google sign in by clicking Button", () => {
    wrapper.find("CustomButton").at(1).simulate("click");
    expect(mockGoogleSignInStart).toHaveBeenCalled();
    expect(mockEmailSignInStart).not.toHaveBeenCalled();
  });
  test("start email sign in by submitting form", async () => {
    await global.actions(wrapper, () =>
      wrapper.find('FormInput[name="email"]').prop<any>("handleChange")({
        target: { value: mockEmail, name: "email" },
      } as any)
    );
    await global.actions(wrapper, () =>
      wrapper.find('FormInput[name="password"]').prop<any>("handleChange")({
        target: { value: mockPassword, name: "password" },
      } as any)
    );
    wrapper.find("form").simulate("submit");
    expect(mockEmailSignInStart).toHaveBeenCalledWith({
      email: mockEmail,
      password: mockPassword,
    });
  });
});
