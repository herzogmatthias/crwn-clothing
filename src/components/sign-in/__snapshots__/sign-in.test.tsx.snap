// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NON CONNECTED SIGN-IN TESTS render SignIn Component 1`] = `
<SignIn
  emailSignInStart={[MockFunction]}
  googleSignInStart={[MockFunction]}
>
  <styled.div>
    <div
      className="sc-AxhUy fTRnYT"
    >
      <styled.h2>
        <h2
          className="sc-AxgMl iLAkrD"
        >
          I already have an account
        </h2>
      </styled.h2>
      <span>
        Sign in with your email and password
      </span>
      <form
        onSubmit={[Function]}
      >
        <FormInput
          handleChange={[Function]}
          label="Email"
          name="email"
          required={true}
          type="email"
          value=""
        >
          <styled.div>
            <div
              className="sc-AxjAm hendVB"
            >
              <styled.input
                name="email"
                onChange={[Function]}
                required={true}
                type="email"
                value=""
              >
                <input
                  className="sc-AxiKw eoQQwa"
                  name="email"
                  onChange={[Function]}
                  required={true}
                  type="email"
                  value=""
                />
              </styled.input>
              <styled.label
                length={0}
              >
                <label
                  className="sc-AxirZ hZnEDN"
                >
                  Email
                </label>
              </styled.label>
            </div>
          </styled.div>
        </FormInput>
        <FormInput
          handleChange={[Function]}
          label="Password"
          name="password"
          required={true}
          type="password"
          value=""
        >
          <styled.div>
            <div
              className="sc-AxjAm hendVB"
            >
              <styled.input
                name="password"
                onChange={[Function]}
                required={true}
                type="password"
                value=""
              >
                <input
                  className="sc-AxiKw kaLtRc"
                  name="password"
                  onChange={[Function]}
                  required={true}
                  type="password"
                  value=""
                />
              </styled.input>
              <styled.label
                length={0}
              >
                <label
                  className="sc-AxirZ hZnEDN"
                >
                  Password
                </label>
              </styled.label>
            </div>
          </styled.div>
        </FormInput>
        <styled.div>
          <div
            className="sc-AxheI eSdjqR"
          >
            <CustomButton
              inverted={false}
              isGoogleSignIn={false}
              type="submit"
            >
              <styled.button
                inverted={false}
                isGoogleSignIn={false}
                type="submit"
              >
                <button
                  className="sc-AxhCb jMBtEZ"
                  type="submit"
                >
                  Sign In
                </button>
              </styled.button>
            </CustomButton>
            <CustomButton
              inverted={false}
              isGoogleSignIn={true}
              onClick={[MockFunction]}
              type="button"
            >
              <styled.button
                inverted={false}
                isGoogleSignIn={true}
                onClick={[MockFunction]}
                type="button"
              >
                <button
                  className="sc-AxhCb gXJAEf"
                  onClick={[MockFunction]}
                  type="button"
                >
                  Sign In with Google
                </button>
              </styled.button>
            </CustomButton>
          </div>
        </styled.div>
      </form>
    </div>
  </styled.div>
</SignIn>
`;
